### ISSUE TITLE
New policy: service account bound to high risk role

### ISSUE THREAD

**Issue by flavio** [kubewarden/kubewarden-controller#902]
Write a Kubewarden policy that can be used to migrate this NeuVector policy:

![image](https://github.com/user-attachments/assets/55f6b78e-8a79-41d5-b93a-16356c972aed)

## Acceptance criteria

- [ ] Define a settings file for the policy
- [ ] Document how the policy should behave

This task is **NOT** about writing the actual policy, we first have to define what the policy is going to do and its UX. Basically this a task about writing the README of the policy

**Comment by jvanz** [comment]
Sorry for the delay. I've paused this task multiple times to do other things. Some relevant links to this issue:

https://github.com/neuvector/neuvector/blob/0e7758c41cc73a607adefd6deceb5358a723d228/controller/resource/kubernetes_rbac.go#L2096
https://open-docs.neuvector.com/policy/admission/#admission-control-rules

The initial README proposal:

---


## High-Risk Service Account Blocker

Kubernetes [service accounts](https://kubernetes.io/docs/concepts/security/service-accounts/) are account types used by workloads running in the cluster. A service account can grant credentials to workloads, enabling them to perform a wide range of cluster operations. This can be dangerous when a workload is able to manipulate resources it is not authorized to. This policy aims to mitigate such risks by preventing resources that utilize high-risk service accounts from being deployed in the cluster. To achieve this, the policy inspects the roles associated with the service account in use. If any of the rules are considered high-risk, the request is rejected.

Every time a resource that defines a service account is submitted to the cluster, the policy will fetch its associated roles and cluster roles and check if their rules are a subset of the rules defined in the policy's configuration. If they are, the resource is rejected.

### Settings

The policy settings consist of a list of rules that service accounts are prohibited from having defined in any of their associated roles and cluster roles. The rules follow the same syntax as those defined in the [roles](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.32/#policyrule-v1-rbac-authorization-k8s-io) specification.

```yaml
blockRules:
# For listing secrets
- apiGroups: [""]
  resources: ["secrets"]
  verbs: ["list"]

# For executing commands in containers
- apiGroups: [""]
  resources: ["pods/exec"]
  verbs: ["create"]

# Full access to all workload resources
- apiGroups: [""]
  resources: ["pods", "pods/log"]
  verbs: ["*"]
- apiGroups: ["apps"]
  resources: ["deployments", "statefulsets", "daemonsets", "replicasets"]
  verbs: ["*"]
- apiGroups: ["batch"]
  resources: ["jobs", "cronjobs"]
  verbs: ["*"]
- apiGroups: ["autoscaling"]
  resources: ["horizontalpodautoscalers"]
  verbs: ["*"]

# Full access to RBAC resources within namespace
- apiGroups: ["rbac.authorization.k8s.io"]
  resources: ["roles", "rolebindings"]
  verbs: ["*"]
```

For example, if the policy is deployed with the previous configuration in a cluster that has a service account like this:

```yaml
---
# ServiceAccount
apiVersion: v1
kind: ServiceAccount
metadata:
  name: super-admin-sa
  namespace: default

---
# Powerful Role with all requested permissions
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: super-admin-role
  namespace: default
rules:
# For listing secrets
- apiGroups: [""]
  resources: ["secrets"]
  verbs: ["list"]

# For executing commands in containers
- apiGroups: [""]
  resources: ["pods/exec"]
  verbs: ["create"]

# Full access to all workload resources
- apiGroups: [""]
  resources: ["pods", "pods/log"]
  verbs: ["*"]
- apiGroups: ["apps"]
  resources: ["deployments", "statefulsets", "daemonsets", "replicasets"]
  verbs: ["*"]
- apiGroups: ["batch"]
  resources: ["jobs", "cronjobs"]
  verbs: ["*"]
- apiGroups: ["autoscaling"]
  resources: ["horizontalpodautoscalers"]
  verbs: ["*"]
# Full access to RBAC resources within namespace
- apiGroups: ["rbac.authorization.k8s.io"]
  resources: ["roles", "rolebindings"]
  verbs: ["*"]

---
# RoleBinding for namespace-scoped permissions
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: super-admin-rolebinding
  namespace: default
subjects:
- kind: ServiceAccount
  name: super-admin-sa
  namespace: default
roleRef:
  kind: Role
  name: super-admin-role
  apiGroup: rbac.authorization.k8s.io
```

The following deployment would not be allowed in the cluster:

```yaml
---
# Deployment using the powerful ServiceAccount
apiVersion: apps/v1
kind: Deployment
metadata:
  name: super-admin-deployment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: super-admin-app
  template:
    metadata:
      labels:
        app: super-admin-app
    spec:
      serviceAccountName: super-admin-sa
      containers:
      - name: main
        image: nginx:alpine
        ports:
        - containerPort: 80
      - name: kubectl
        image: bitnami/kubectl:latest
        command: ["sleep", "infinity"]
```

**Comment by flavio** [comment]
I think the policy description looks good. We have to take a smart approach at solving this problem, the policy **must not** do context aware calls to get ServiceAccount, RoleBinding,... from the Kubernetes API. I don't want us to make tons of requests (it's slow) and I don't want us to reinvent the wheel to check what is allowed and denied (on top of redoing Kubernetes' work, we risk to introduce bugs).

Kubernetes' has a dedicated API that can be used to check if a user/ServiceAccount can perform a certain operation. We **must** use that.

I've thrown together [this demo](https://github.com/flavio/kube-rs-can-i-do) that shows the Rust code that makes use of this API.

To summarize, this is what we should be doing:

- Kubewarden waPC: introduce a new host capability: `kubernetes/can-i`
- policy-evaluator: implement the new host capability
- policy-server: consume the new version of policy-evaluator
- kwctl: consume the new version of policy-evaluator
- helm chart: update the SA used by the policy server, it must have the RBAC permissions to create `SubjectAccessReview` objects
- document all the things

Once this is done, we can start to implement this policy.

Congratulations, this just turned into an epic! 🥚 🐣 🐥 🐔 😄

**Comment by jvanz** [comment]
Yes, that's look a good approach to me. I just added the issue to the milestone.

**Comment by viccuad** [comment]
Implementing this new [`can_i` capability on the new kubewarden-sdk-js](https://github.com/kubewarden/policy-sdk-js/issues/116) will be done as part of the GsoC tasks, hence this card is closed as done.
